File: src/main/scala/org/openapitools/client/api/OffersApi.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import java.io.File
import java.net.URLEncoder
import java.util.UUID

import org.http4s._
import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.http4s.client._
import org.http4s.client.blaze.PooledHttp1Client
import org.http4s.headers._

import org.joda.time.DateTime

import scalaz.concurrent.Task

import HelperCodecs._

import org.openapitools.client.api.ApiClientDataErrorResponse
import org.openapitools.client.api.ApiForbiddenErrorResponse
import org.openapitools.client.api.ApiLimitErrorResponse
import org.openapitools.client.api.ApiLockedErrorResponse
import org.openapitools.client.api.ApiNotFoundErrorResponse
import org.openapitools.client.api.ApiServerErrorResponse
import org.openapitools.client.api.ApiUnauthorizedErrorResponse
import org.openapitools.client.api.CurrencyType
import org.openapitools.client.api.DeleteCampaignOffersRequest
import org.openapitools.client.api.DeleteCampaignOffersResponse
import org.openapitools.client.api.EmptyApiResponse
import org.openapitools.client.api.GetAllOffersResponse
import org.openapitools.client.api.GetCampaignOffersRequest
import org.openapitools.client.api.GetCampaignOffersResponse
import org.openapitools.client.api.GetOfferRecommendationsRequest
import org.openapitools.client.api.GetOfferRecommendationsResponse
import org.openapitools.client.api.GetOffersResponse
import org.openapitools.client.api.UpdateCampaignOffersRequest

object OffersApi {

  val client = PooledHttp1Client()

  def escape(value: String): String = URLEncoder.encode(value, "utf-8").replaceAll("\\+", "%20")

  def deleteCampaignOffers(host: String, campaignId: Long, deleteCampaignOffersRequest: DeleteCampaignOffersRequest): Task[DeleteCampaignOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[DeleteCampaignOffersResponse] = jsonOf[DeleteCampaignOffersResponse]

    val path = "/campaigns/{campaignId}/offers/delete".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      )

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(host + path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(deleteCampaignOffersRequest)
      resp          <- client.expect[DeleteCampaignOffersResponse](req)

    } yield resp
  }

  def getAllOffers(host: String, campaignId: Long, feedId: Long, chunk: Integer)(implicit feedIdQuery: QueryParam[Long], chunkQuery: QueryParam[Integer]): Task[GetAllOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetAllOffersResponse] = jsonOf[GetAllOffersResponse]

    val path = "/campaigns/{campaignId}/offers/all".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.GET
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("feedId", Some(feedIdQuery.toParamString(feedId))), ("chunk", Some(chunkQuery.toParamString(chunk))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(host + path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType))
      resp          <- client.expect[GetAllOffersResponse](req)

    } yield resp
  }

  def getCampaignOffers(host: String, campaignId: Long, getCampaignOffersRequest: GetCampaignOffersRequest, pageToken: String, limit: Integer)(implicit pageTokenQuery: QueryParam[String], limitQuery: QueryParam[Integer]): Task[GetCampaignOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetCampaignOffersResponse] = jsonOf[GetCampaignOffersResponse]

    val path = "/campaigns/{campaignId}/offers".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("pageToken", Some(page_tokenQuery.toParamString(page_token))), ("limit", Some(limitQuery.toParamString(limit))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(host + path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(getCampaignOffersRequest)
      resp          <- client.expect[GetCampaignOffersResponse](req)

    } yield resp
  }

  def getOfferRecommendations(host: String, businessId: Long, getOfferRecommendationsRequest: GetOfferRecommendationsRequest, pageToken: String, limit: Integer)(implicit pageTokenQuery: QueryParam[String], limitQuery: QueryParam[Integer]): Task[GetOfferRecommendationsResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetOfferRecommendationsResponse] = jsonOf[GetOfferRecommendationsResponse]

    val path = "/businesses/{businessId}/offers/recommendations".replaceAll("\\{" + "businessId" + "\\}",escape(businessId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("pageToken", Some(page_tokenQuery.toParamString(page_token))), ("limit", Some(limitQuery.toParamString(limit))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(host + path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(getOfferRecommendationsRequest)
      resp          <- client.expect[GetOfferRecommendationsResponse](req)

    } yield resp
  }

  def getOffers(host: String, campaignId: Long, query: String, feedId: Long, shopCategoryId: String, currency: CurrencyType, matched: Boolean, page: Integer = 1, pageSize: Integer)(implicit queryQuery: QueryParam[String], feedIdQuery: QueryParam[Long], shopCategoryIdQuery: QueryParam[String], currencyQuery: QueryParam[CurrencyType], matchedQuery: QueryParam[Boolean], pageQuery: QueryParam[Integer], pageSizeQuery: QueryParam[Integer]): Task[GetOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetOffersResponse] = jsonOf[GetOffersResponse]

    val path = "/campaigns/{campaignId}/offers".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.GET
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("query", Some(queryQuery.toParamString(query))), ("feedId", Some(feedIdQuery.toParamString(feedId))), ("shopCategoryId", Some(shopCategoryIdQuery.toParamString(shopCategoryId))), ("currency", Some(currencyQuery.toParamString(currency))), ("matched", Some(matchedQuery.toParamString(matched))), ("page", Some(pageQuery.toParamString(page))), ("pageSize", Some(pageSizeQuery.toParamString(pageSize))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(host + path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType))
      resp          <- client.expect[GetOffersResponse](req)

    } yield resp
  }

  def updateCampaignOffers(host: String, campaignId: Long, updateCampaignOffersRequest: UpdateCampaignOffersRequest): Task[EmptyApiResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[EmptyApiResponse] = jsonOf[EmptyApiResponse]

    val path = "/campaigns/{campaignId}/offers/update".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      )

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(host + path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(updateCampaignOffersRequest)
      resp          <- client.expect[EmptyApiResponse](req)

    } yield resp
  }

}

class HttpServiceOffersApi(service: HttpService) {
  val client = Client.fromHttpService(service)

  def escape(value: String): String = URLEncoder.encode(value, "utf-8").replaceAll("\\+", "%20")

  def deleteCampaignOffers(campaignId: Long, deleteCampaignOffersRequest: DeleteCampaignOffersRequest): Task[DeleteCampaignOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[DeleteCampaignOffersResponse] = jsonOf[DeleteCampaignOffersResponse]

    val path = "/campaigns/{campaignId}/offers/delete".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      )

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(deleteCampaignOffersRequest)
      resp          <- client.expect[DeleteCampaignOffersResponse](req)

    } yield resp
  }

  def getAllOffers(campaignId: Long, feedId: Long, chunk: Integer)(implicit feedIdQuery: QueryParam[Long], chunkQuery: QueryParam[Integer]): Task[GetAllOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetAllOffersResponse] = jsonOf[GetAllOffersResponse]

    val path = "/campaigns/{campaignId}/offers/all".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.GET
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("feedId", Some(feedIdQuery.toParamString(feedId))), ("chunk", Some(chunkQuery.toParamString(chunk))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType))
      resp          <- client.expect[GetAllOffersResponse](req)

    } yield resp
  }

  def getCampaignOffers(campaignId: Long, getCampaignOffersRequest: GetCampaignOffersRequest, pageToken: String, limit: Integer)(implicit pageTokenQuery: QueryParam[String], limitQuery: QueryParam[Integer]): Task[GetCampaignOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetCampaignOffersResponse] = jsonOf[GetCampaignOffersResponse]

    val path = "/campaigns/{campaignId}/offers".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("pageToken", Some(page_tokenQuery.toParamString(page_token))), ("limit", Some(limitQuery.toParamString(limit))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(getCampaignOffersRequest)
      resp          <- client.expect[GetCampaignOffersResponse](req)

    } yield resp
  }

  def getOfferRecommendations(businessId: Long, getOfferRecommendationsRequest: GetOfferRecommendationsRequest, pageToken: String, limit: Integer)(implicit pageTokenQuery: QueryParam[String], limitQuery: QueryParam[Integer]): Task[GetOfferRecommendationsResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetOfferRecommendationsResponse] = jsonOf[GetOfferRecommendationsResponse]

    val path = "/businesses/{businessId}/offers/recommendations".replaceAll("\\{" + "businessId" + "\\}",escape(businessId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("pageToken", Some(page_tokenQuery.toParamString(page_token))), ("limit", Some(limitQuery.toParamString(limit))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(getOfferRecommendationsRequest)
      resp          <- client.expect[GetOfferRecommendationsResponse](req)

    } yield resp
  }

  def getOffers(campaignId: Long, query: String, feedId: Long, shopCategoryId: String, currency: CurrencyType, matched: Boolean, page: Integer = 1, pageSize: Integer)(implicit queryQuery: QueryParam[String], feedIdQuery: QueryParam[Long], shopCategoryIdQuery: QueryParam[String], currencyQuery: QueryParam[CurrencyType], matchedQuery: QueryParam[Boolean], pageQuery: QueryParam[Integer], pageSizeQuery: QueryParam[Integer]): Task[GetOffersResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[GetOffersResponse] = jsonOf[GetOffersResponse]

    val path = "/campaigns/{campaignId}/offers".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.GET
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      ("query", Some(queryQuery.toParamString(query))), ("feedId", Some(feedIdQuery.toParamString(feedId))), ("shopCategoryId", Some(shopCategoryIdQuery.toParamString(shopCategoryId))), ("currency", Some(currencyQuery.toParamString(currency))), ("matched", Some(matchedQuery.toParamString(matched))), ("page", Some(pageQuery.toParamString(page))), ("pageSize", Some(pageSizeQuery.toParamString(pageSize))))

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType))
      resp          <- client.expect[GetOffersResponse](req)

    } yield resp
  }

  def updateCampaignOffers(campaignId: Long, updateCampaignOffersRequest: UpdateCampaignOffersRequest): Task[EmptyApiResponse] = {
    implicit val returnTypeDecoder: EntityDecoder[EmptyApiResponse] = jsonOf[EmptyApiResponse]

    val path = "/campaigns/{campaignId}/offers/update".replaceAll("\\{" + "campaignId" + "\\}",escape(campaignId.toString))

    val httpMethod = Method.POST
    val contentType = `Content-Type`(MediaType.`application/json`)
    val headers = Headers(
      )
    val queryParams = Query(
      )

    for {
      uri           <- Task.fromDisjunction(Uri.fromString(path))
      uriWithParams =  uri.copy(query = queryParams)
      req           =  Request(method = httpMethod, uri = uriWithParams, headers = headers.put(contentType)).withBody(updateCampaignOffersRequest)
      resp          <- client.expect[EmptyApiResponse](req)

    } yield resp
  }

}

File: lib/service/OffersApi.cpp
#include "OffersApi.h"

using namespace Tiny;



        Response<
            DeleteCampaignOffersResponse
        >
        OffersApi::
        deleteCampaignOffers(
            
            long campaignId
            , 
            
            DeleteCampaignOffersRequest deleteCampaignOffersRequest
            
        )
        {
            std::string url = basepath + "/campaigns/{campaignId}/offers/delete"; //campaignId 


            // Headers  | 

            // Query    | 

            // Form     | 
            addHeader("Content-Type", "application/json");



                std::string s_campaignId("{");
                s_campaignId.append("campaignId");
                s_campaignId.append("}");

                int pos = url.find(s_campaignId);

                url.erase(pos, s_campaignId.length());
                url.insert(pos, stringify(campaignId));


            std::string payload = "";
            // Send Request
            // METHOD | POST
            // Body     | deleteCampaignOffersRequest



            payload = deleteCampaignOffersRequest.toJson().dump();

            int httpCode = sendRequest(url, "POST", reinterpret_cast<uint8_t*>(&payload[0]), payload.length());

            // Handle Request
            String output = getResponseBody();
            std::string output_string = output.c_str();




            DeleteCampaignOffersResponse obj(output_string);


            Response<DeleteCampaignOffersResponse> response(obj, httpCode);
            return response;
        }

        Response<
            GetAllOffersResponse
        >
        OffersApi::
        getAllOffers(
            
            long campaignId
            , 
            
            long feedId
            , 
            
            int chunk
            
        )
        {
            std::string url = basepath + "/campaigns/{campaignId}/offers/all"; //campaignId 


            // Headers  | 

            // Query    | feedId chunk 
            addQueryParam("feedId",feedId);
            addQueryParam("chunk",chunk);

            // Form     | 



                std::string s_campaignId("{");
                s_campaignId.append("campaignId");
                s_campaignId.append("}");

                int pos = url.find(s_campaignId);

                url.erase(pos, s_campaignId.length());
                url.insert(pos, stringify(campaignId));


            std::string payload = "";
            // Send Request
            // METHOD | GET
            // Body     | 
            int httpCode = sendRequest(url, "GET", reinterpret_cast<uint8_t*>(&payload[0]), payload.length());

            // Handle Request
            String output = getResponseBody();
            std::string output_string = output.c_str();




            GetAllOffersResponse obj(output_string);


            Response<GetAllOffersResponse> response(obj, httpCode);
            return response;
        }

        Response<
            GetCampaignOffersResponse
        >
        OffersApi::
        getCampaignOffers(
            
            long campaignId
            , 
            
            GetCampaignOffersRequest getCampaignOffersRequest
            , 
            
            std::string pageToken
            , 
            
            int limit
            
        )
        {
            std::string url = basepath + "/campaigns/{campaignId}/offers"; //campaignId 


            // Headers  | 

            // Query    | pageToken limit 
            addQueryParam("page_token",pageToken);
            addQueryParam("limit",limit);

            // Form     | 
            addHeader("Content-Type", "application/json");



                std::string s_campaignId("{");
                s_campaignId.append("campaignId");
                s_campaignId.append("}");

                int pos = url.find(s_campaignId);

                url.erase(pos, s_campaignId.length());
                url.insert(pos, stringify(campaignId));


            std::string payload = "";
            // Send Request
            // METHOD | POST
            // Body     | getCampaignOffersRequest



            payload = getCampaignOffersRequest.toJson().dump();

            int httpCode = sendRequest(url, "POST", reinterpret_cast<uint8_t*>(&payload[0]), payload.length());

            // Handle Request
            String output = getResponseBody();
            std::string output_string = output.c_str();




            GetCampaignOffersResponse obj(output_string);


            Response<GetCampaignOffersResponse> response(obj, httpCode);
            return response;
        }

        Response<
            GetOfferRecommendationsResponse
        >
        OffersApi::
        getOfferRecommendations(
            
            long businessId
            , 
            
            GetOfferRecommendationsRequest getOfferRecommendationsRequest
            , 
            
            std::string pageToken
            , 
            
            int limit
            
        )
        {
            std::string url = basepath + "/businesses/{businessId}/offers/recommendations"; //businessId 


            // Headers  | 

            // Query    | pageToken limit 
            addQueryParam("page_token",pageToken);
            addQueryParam("limit",limit);

            // Form     | 
            addHeader("Content-Type", "application/json");



                std::string s_businessId("{");
                s_businessId.append("businessId");
                s_businessId.append("}");

                int pos = url.find(s_businessId);

                url.erase(pos, s_businessId.length());
                url.insert(pos, stringify(businessId));


            std::string payload = "";
            // Send Request
            // METHOD | POST
            // Body     | getOfferRecommendationsRequest



            payload = getOfferRecommendationsRequest.toJson().dump();

            int httpCode = sendRequest(url, "POST", reinterpret_cast<uint8_t*>(&payload[0]), payload.length());

            // Handle Request
            String output = getResponseBody();
            std::string output_string = output.c_str();




            GetOfferRecommendationsResponse obj(output_string);


            Response<GetOfferRecommendationsResponse> response(obj, httpCode);
            return response;
        }

        Response<
            GetOffersResponse
        >
        OffersApi::
        getOffers(
            
            long campaignId
            , 
            
            std::string query
            , 
            
            long feedId
            , 
            
            std::string shopCategoryId
            , 
            
            CurrencyType currency
            , 
            
            bool matched
            , 
            
            int page
            , 
            
            int pageSize
            
        )
        {
            std::string url = basepath + "/campaigns/{campaignId}/offers"; //campaignId 


            // Headers  | 

            // Query    | query feedId shopCategoryId currency matched page pageSize 
            addQueryParam("query",query);
            addQueryParam("feedId",feedId);
            addQueryParam("shopCategoryId",shopCategoryId);
            addQueryParam("currency",currency);
            addQueryParam("matched",matched);
            addQueryParam("page",page);
            addQueryParam("pageSize",pageSize);

            // Form     | 



                std::string s_campaignId("{");
                s_campaignId.append("campaignId");
                s_campaignId.append("}");

                int pos = url.find(s_campaignId);

                url.erase(pos, s_campaignId.length());
                url.insert(pos, stringify(campaignId));


            std::string payload = "";
            // Send Request
            // METHOD | GET
            // Body     | 
            int httpCode = sendRequest(url, "GET", reinterpret_cast<uint8_t*>(&payload[0]), payload.length());

            // Handle Request
            String output = getResponseBody();
            std::string output_string = output.c_str();




            GetOffersResponse obj(output_string);


            Response<GetOffersResponse> response(obj, httpCode);
            return response;
        }

        Response<
            EmptyApiResponse
        >
        OffersApi::
        updateCampaignOffers(
            
            long campaignId
            , 
            
            UpdateCampaignOffersRequest updateCampaignOffersRequest
            
        )
        {
            std::string url = basepath + "/campaigns/{campaignId}/offers/update"; //campaignId 


            // Headers  | 

            // Query    | 

            // Form     | 
            addHeader("Content-Type", "application/json");



                std::string s_campaignId("{");
                s_campaignId.append("campaignId");
                s_campaignId.append("}");

                int pos = url.find(s_campaignId);

                url.erase(pos, s_campaignId.length());
                url.insert(pos, stringify(campaignId));


            std::string payload = "";
            // Send Request
            // METHOD | POST
            // Body     | updateCampaignOffersRequest



            payload = updateCampaignOffersRequest.toJson().dump();

            int httpCode = sendRequest(url, "POST", reinterpret_cast<uint8_t*>(&payload[0]), payload.length());

            // Handle Request
            String output = getResponseBody();
            std::string output_string = output.c_str();




            EmptyApiResponse obj(output_string);


            Response<EmptyApiResponse> response(obj, httpCode);
            return response;
        }



